File: studio/backend/api/v1/schemas/reports.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

AllowedReportType = Literal[
    "market_analysis",
    "competitor_overview",
    "trend_report",
    "industry_deep_dive",
]


class ReportCreate(BaseModel):
    topic: str = Field(min_length=10, max_length=500)
    report_type: AllowedReportType


class ReportCreateResponse(BaseModel):
    report_id: UUID
    job_id: UUID
    celery_task_id: str
    status: str


class ReportListItem(BaseModel):
    id: UUID
    title: str
    topic: str
    report_type: str
    status: str
    created_at: datetime
    completed_at: datetime | None = None


class ReportDetail(ReportListItem):
    content_md: str | None = None
    word_count: int | None = None


class ReportListResponse(BaseModel):
    items: list[ReportListItem]
    total: int
    page: int
    page_size: int

File: studio/frontend/src/lib/api/reports.ts
import { apiRequest, buildApiUrl, type ApiError } from "@/lib/api/client";

export type ReportStatus = "pending" | "running" | "done" | "failed";

export type ReportType =
  | "market_analysis"
  | "competitor_overview"
  | "trend_report"
  | "industry_deep_dive";

export interface Report {
  id: string;
  title: string;
  topic: string;
  report_type: ReportType;
  status: ReportStatus;
  content_md: string | null;
  word_count: number | null;
  created_at: string;
  completed_at: string | null;
  job_id?: string | null;
}

export interface PaginatedReports {
  items: Report[];
  total: number;
  page: number;
  page_size: number;
}

export const getReports = async (
  token: string,
  page = 1,
  pageSize = 10,
): Promise<PaginatedReports> =>
  apiRequest<PaginatedReports>(
    `/reports?page=${page}&page_size=${pageSize}`,
    token,
  );

export const getReport = async (
  token: string,
  reportId: string,
): Promise<Report> =>
  apiRequest<Report>(`/reports/${reportId}`, token);

export const deleteReport = async (
  token: string,
  reportId: string,
): Promise<void> => {
  await apiRequest<void>(`/reports/${reportId}`, token, {
    method: "DELETE",
  });
};

export const getPdfUrl = (reportId: string): string =>
  buildApiUrl(`/reports/${reportId}/export/pdf`);

export type { ApiError };
